File: app/typescript/src/application/service/InventoryService.ts
import { InventoryUseCase, InventoryQuery } from '../port/in/InventoryUseCase';
import { InventoryRepository } from '../port/out/InventoryRepository';
import { Inventory } from '../../domain/model/inventory/Inventory';

/**
 * 在庫 Application Service
 * Use Caseを実装し、在庫照会のビジネスロジックを調整
 */
export class InventoryService implements InventoryUseCase {
  constructor(private readonly inventoryRepository: InventoryRepository) {}

  async getInventory(locationCode: string, itemCode: string): Promise<Inventory | null> {
    return await this.inventoryRepository.findByLocationAndItem(locationCode, itemCode);
  }

  async getInventoriesByLocation(locationCode: string): Promise<Inventory[]> {
    return await this.inventoryRepository.findByLocation(locationCode);
  }

  async getInventoriesByItem(itemCode: string): Promise<Inventory[]> {
    return await this.inventoryRepository.findByItem(itemCode);
  }

  async queryInventories(query: InventoryQuery): Promise<Inventory[]> {
    if (query.場所コード && query.品目コード) {
      const inventory = await this.inventoryRepository.findByLocationAndItem(
        query.場所コード,
        query.品目コード
      );
      return inventory ? [inventory] : [];
    }

    if (query.場所コード) {
      return await this.inventoryRepository.findByLocation(query.場所コード);
    }

    if (query.品目コード) {
      return await this.inventoryRepository.findByItem(query.品目コード);
    }

    return await this.inventoryRepository.findAll();
  }
}

File: app/java/src/main/java/com/example/production/application/service/InventoryService.java
package com.example.production.application.service;

import com.example.production.application.port.in.command.StockChangeCommand;
import com.example.production.application.port.in.command.StockStatusChangeCommand;
import com.example.production.application.port.out.StockRepository;
import com.example.production.domain.model.inventory.InsufficientStockException;
import com.example.production.domain.model.inventory.Stock;
import com.example.production.domain.model.inventory.StockStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

/**
 * 在庫管理サービス
 */
@Service
@RequiredArgsConstructor
public class InventoryService {

    private final StockRepository stockRepository;

    /**
     * 在庫を取得する
     */
    public Stock getStock(String locationCode, String itemCode) {
        return stockRepository.findByLocationAndItem(locationCode, itemCode)
                .orElse(Stock.empty(locationCode, itemCode));
    }

    /**
     * 場所別の在庫一覧を取得する
     */
    public List<Stock> getStocksByLocation(String locationCode) {
        return stockRepository.findByLocationCode(locationCode);
    }

    /**
     * 品目別の在庫一覧を取得する
     */
    public List<Stock> getStocksByItem(String itemCode) {
        return stockRepository.findByItemCode(itemCode);
    }

    /**
     * 在庫を増加する
     */
    @Transactional
    public void increaseStock(StockChangeCommand command) {
        Stock existingStock = stockRepository.findByLocationAndItem(command.getLocationCode(), command.getItemCode())
                .orElse(null);

        if (existingStock == null) {
            // 新規作成
            Stock newStock = Stock.builder()
                    .locationCode(command.getLocationCode())
                    .itemCode(command.getItemCode())
                    .stockQuantity(command.getQuantity())
                    .passedQuantity(command.getStockStatus() == StockStatus.PASSED ? command.getQuantity() : BigDecimal.ZERO)
                    .defectiveQuantity(command.getStockStatus() == StockStatus.DEFECTIVE ? command.getQuantity() : BigDecimal.ZERO)
                    .uninspectedQuantity(command.getStockStatus() == StockStatus.UNINSPECTED ? command.getQuantity() : BigDecimal.ZERO)
                    .build();
            stockRepository.save(newStock);
        } else {
            // 更新
            stockRepository.increaseByStatus(
                    command.getLocationCode(),
                    command.getItemCode(),
                    command.getQuantity(),
                    command.getStockStatus()
            );
        }
    }

    /**
     * 在庫を減少する
     */
    @Transactional
    public void decreaseStock(StockChangeCommand command) {
        Stock stock = getStock(command.getLocationCode(), command.getItemCode());
        BigDecimal currentQuantity = stock.getQuantityByStatus(command.getStockStatus());

        if (currentQuantity.compareTo(command.getQuantity()) < 0) {
            throw new InsufficientStockException(
                    String.format("在庫が不足しています。場所: %s, 品目: %s, 要求数: %s, 現在数: %s",
                            command.getLocationCode(), command.getItemCode(),
                            command.getQuantity(), currentQuantity)
            );
        }

        stockRepository.decreaseByStatus(
                command.getLocationCode(),
                command.getItemCode(),
                command.getQuantity(),
                command.getStockStatus()
        );
    }

    /**
     * 在庫状態を変更する
     */
    @Transactional
    public void changeStockStatus(StockStatusChangeCommand command) {
        Stock stock = getStock(command.getLocationCode(), command.getItemCode());
        BigDecimal currentQuantity = stock.getQuantityByStatus(command.getFromStatus());

        if (currentQuantity.compareTo(command.getQuantity()) < 0) {
            throw new InsufficientStockException(
                    String.format("%sの在庫が不足しています。場所: %s, 品目: %s, 要求数: %s, 現在数: %s",
                            command.getFromStatus().getDisplayName(),
                            command.getLocationCode(), command.getItemCode(),
                            command.getQuantity(), currentQuantity)
            );
        }

        stockRepository.changeStatus(
                command.getLocationCode(),
                command.getItemCode(),
                command.getQuantity(),
                command.getFromStatus(),
                command.getToStatus()
        );
    }
}
